crates/bindings/src/lib.rs

```diff
@@ -235,204 +233,77 @@ pub fn delete_by_rel(table_id: TableId, relation: &[impl Serialize]) -> Result<u
-        self.reader.pos.set(0);
-        if self.inner.read(&mut self.reader.buf) == 0 {
-            return None;
-        }
-        Some(self.decode())
+impl std::ops::Deref for IterBuf {
+    type Target = Vec<u8>;
+    fn deref(&self) -> &Self::Target {
+        &self.buf
     }
 }
-
-/// Describe a named index with an index type over a set of columns identified by their IDs.
-#[derive(Clone, Copy)]
-pub struct IndexDesc<'a> {
-    /// The name of the index.
-    pub name: &'a str,
-    /// The type of index used, i.e. the strategy used for indexing.
-    pub ty: IndexType,
-    /// The set of columns indexed over given by the identifiers of the columns.
-    pub col_ids: &'a [u16],
-}
-
-/// A trait for the set of types serializable, deserializable, and convertible to `AlgebraicType`.
-///
-/// Additionally, the type knows its own table name, its column attributes, and indices.
-pub trait TableType: SpacetimeType + DeserializeOwned + Serialize {
-    const TABLE_NAME: &'static str;
-    const TABLE_ACCESS: StAccess;
-    const COLUMN_ATTRS: &'static [ColumnAttribute];
-    const INDEXES: &'static [IndexDesc<'static>];
-    const SCHEDULED_REDUCER_NAME: Option<&'static str> = None;
-
-    type InsertResult: sealed::InsertResult<T = Self>;
-
-    /// Returns the ID of this table.
-    fn table_id() -> TableId;
-
-    /// Insert `ins` as a row in this table.
-    fn insert(ins: Self) -> Self::InsertResult {
-        insert(Self::table_id(), ins)
-    }
-
-    /// Returns an iterator over the rows in this table.
-    fn iter() -> TableIter<Self> {
-        table_iter(Self::table_id()).unwrap()
-    }
-
-    /// Returns an iterator filtered by `filter` over the rows in this table.
-    ///
-    /// **NOTE:** Do not use directly. This is exposed as `query!(...)`.
-    #[doc(hidden)]
-    fn iter_filtered(filter: spacetimedb_lib::filter::Expr) -> TableIter<Self> {
-        table_iter_filtered(Self::table_id(), &filter).unwrap()
-    }
-
-    /// Deletes this row `self` from the table.
-    ///
-    /// Returns `true` if the row was deleted.
-    fn delete(&self) -> bool {
-        let count = delete_by_rel(Self::table_id(), from_ref(self)).unwrap();
-        debug_assert!(count < 2);
-        count == 1
+impl std::ops::DerefMut for IterBuf {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.buf
     }
 }
 
-mod sealed {
-    use super::*;
-
-    /// A trait of result types which know how to convert a `Result<T: TableType>` into itself.
-    pub trait InsertResult {
-        type T: TableType;
-        fn from_res(res: Result<Self::T>) -> Self;
-    }
-}
+#[doc(hidden)]
+pub trait Captures<'a> {}
+impl<'a, T> Captures<'a> for T {}
 
 /// A UNIQUE constraint violation on table type `T` was attempted.
-pub struct UniqueConstraintViolation<T: TableType>(PhantomData<T>);
-impl<T: TableType> fmt::Debug for UniqueConstraintViolation<T> {
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct UniqueConstraintViolation;
```

Comment:
This could potentially contain `table: String, column: String` for example. That can be future work, but please leave a TODO comment in here to get better error messages.